src/firefly/js/ui/panel/TabPanel.jsx

```diff
@@ -311,8 +209,146 @@ FieldGroupTabs.propTypes = {
+ Exported function and components
+----------------------------------------------------------------------------------------------*/
+
+/**
+ * For backward compatibility, this is a composite of JoyUI's Tab and TabPanel
+ * 'label' is converted to Tab, and 'children' are wrapped inside a Joy TabPanel
+ * It is designed to be used with TabPanel.
+ */
+export const Tab = React.memo( ({label, name, value, startDecorator, removable, onTabRemove}) => {
+    // maxTitleWidth:  deprecated.  default to 400
+    return;
+});
+
+Tab.propTypes = {
+    label:  PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
+    name: PropTypes.string,
+    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+    startDecorator: PropTypes.node,
+    removable: PropTypes.bool,
+    onTabRemove: PropTypes.func,
+};
+
+
+export function switchTab(componentKey, selectedIdx) {
+    dispatchComponentStateChange(componentKey, {selectedIdx});
+}
+
+
+/*---------------------------------------------------------------------------------------------
+ Internal use only
+----------------------------------------------------------------------------------------------*/
+
+const ResizeTabHeader = sizeMe({refreshRate: 16})(TabHeader);
+
+function TabHeader({children, slotProps}) {
+    const toolbarEl = useRef(null);
+    const [maxTitleWidth, setMaxTitleWidth] = useState(400);
+
+
+    const childrenAry = React.Children.toArray(children);
+    // get the headers(Tab)
+    const tabHeaders = childrenAry.map((c, idx) => getHeaderFromTab(c.props, idx, maxTitleWidth));
+
+    const tabListVar = slotProps?.tabList?.variant || 'soft';
+    const activeBg = tabListVar === 'plain' ? 'neutral.softBg' : 'background.surface';
+
+    useEffect(() => {
+        const tbEl = toolbarEl.current;
+        if (tbEl) {
+            const width = tbEl.getBoundingClientRect().width;
+            setMaxTitleWidth( (width/tabHeaders.length) -4 );   // -4 account for margin and padding
+        }
+    }, [toolbarEl?.current?.getBoundingClientRect()?.width]);
+
+    return (
+        <TabList
+            ref={toolbarEl}
+            sx={{
+                flexGrow: 1,
+                overflow: 'auto',
+                scrollSnapType: 'x mandatory',
+                '&::-webkit-scrollbar': { display: 'none' },
+                [`& .${tabClasses.root}`]: {
+                    '&[aria-selected="true"]': {
+                        bgcolor: activeBg,
+                        borderColor: 'divider',
+                        '&::before': {
+                            content: '""',
+                            display: 'block',
+                            position: 'absolute',
+                            height: 2,
+                            bottom: -2,
+                            left: 0,
+                            right: 0,
+                            bgcolor: activeBg,
+                        },
```

Comment:
You might want to add some documentation here. I have no idea what this is doing.